### ISSUE TITLE
pattern/expression consistency in array/object spacing rules

### ISSUE THREAD

**Issue by benderTheCrime** [eslint/eslint#8727]
**Tell us about your environment**:

* **ESLint Version:**
4.0.0

* **Node Version:**
8.1.0

* **npm Version:**
5.0.3

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint@^7.2.3

**Please show your full configuration:**
https://github.com/benderTheCrime/eslint-plugin-benderthecrime/blob/develop/.eslintrc.json

**What did you do? Please include the actual source code causing the issue.**

```js
const arr = ['foo', 'bar'];
const [foo, bar] = arr;

function baz([foo, bar]) {}
```

**What did you expect to happen?**
There should be no array-element-newline error on the destructure assignments, but there should be an error on the variable initialization.

**What actually happened? Please include the actual, raw output from ESLint.**:
There were eslint errors present on both ArrayPatterns and ArrayExpressions:

```bash
  38:26  error    There should be no linebreak after '['            array-bracket-newline
  40:17  error    There should be no linebreak before ']'           array-bracket-newline
  47:26  error    There should be no linebreak after '['            array-bracket-newline
  49:17  error    There should be no linebreak before ']'           array-bracket-newline
```

In the object-curly-newline rule there exist configuration options specific to both ObjectPatterns and ObjectExpressions allowing you to configure the multiline behavior of the object curlies on ObjectPattern and ObjectExpression independently.

This issue is to raise three asks/possibilites:
- Expose ArrayPattern/ArrayExpression configuration options in the configuration object associated with array-bracket-newline
- Expose ArrayPattern/ArrayExpression configuration options in the configuration object associated with array-element-newline
- Expose ObjectPattern/ObjectExpression configuration options in the configuration object associated with object-property-newline

Granted each of these rules has different configuration options, but I think the paradigm of exposing the actual node type definitions at the top of the configuration object is enough for the sake of consistency here.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
